app/src/main/java/com/sdpteam/connectout/post/model/PostFirebaseDataSource.java

```diff
@@ -1,63 +1,169 @@
+            }
+            final Post post = Objects.requireNonNull(postResult.value());
+
+            if (post.getVisibility() == null) {
+                result.complete(new Result<>(null, false, "Event has visibility set to NULL which is not expected"));
+            } else if (post.getVisibility().equals(PUBLIC)) {
+                result.complete(new Result<>(post, true));
+            } else if (post.getVisibility().equals(SEMIPRIVATE)) {
+                final Event event = (new EventFirebaseDataSource().getEvent(post.getEventId())).join();
+                if (event != null) {
+                    if (event.getOrganizer().equals(post.getProfileId()) || event.getParticipants().contains(post.getProfileId())) {
+                        result.complete(new Result<>(post, true));
+                    } else {
+                        result.complete(new Result<>(null, false, "Error occurred, user has not access to this post"));
+                    }
+                } else {
+                    result.complete(new Result<>(null, false,
+                            "Error the event associated to the post does not exist! (maybe a timeout issue) postId" + postId + " eventId" + post.getEventId()));
+                }
+            }
+        }).join();
+        return result;
     }
 
-    //TODO fix through firebase task
     @Override
     public CompletableFuture<Result<List<Post>>> fetchAllPosts(String userId) {
-        CompletableFuture<Result<List<Post>>> future = new CompletableFuture<>();
-        FirebaseDatabase.getInstance().getReference().child(POSTS)
-                .orderByKey()
-                .get()
-                .addOnCompleteListener(t -> {
-
-                    List<Post> posts = new ArrayList<>();
-                    for (DataSnapshot child : t.getResult().getChildren()) {
-                        Post post = child.getValue(Post.class);
-                        posts.add(post);
-                    }
-                    future.complete(new Result<>(posts, true, ""));
-                });
+        return eventIdsUserCanAccess(userId).thenCompose(allEventsUserCanAccess -> {
+            final CompletableFuture<Result<List<Post>>> result = new CompletableFuture<>();
 
-        return future;
+            FirebaseDatabase.getInstance().getReference().child(POSTS).get().addOnCompleteListener(task -> {
+                if (task.isSuccessful() && task.getResult().exists() && !forceFail) {
+                    final DataSnapshot dataSnapshot = task.getResult();
+                    final List<Post> postList = StreamSupport.stream(dataSnapshot.getChildren().spliterator(), false)
+                            .map(childSnapshot -> childSnapshot.getValue(Post.class))
+                            .filter(Objects::nonNull)
+                            .filter(post -> {
+                                if (post.getVisibility() != null && post.getVisibility().equals(SEMIPRIVATE) && post.getEventId() != null) {
```

Comment:
Can you refactor this condition as a method on post directly and give it a cool name